verification/certifyKey_test.go

```diff
@@ -3,61 +3,521 @@
+		if len(cert.UnknownExtKeyUsage) > 0 {
+			unknownKeyUsages := []string{}
+			for _, eku := range cert.UnknownExtKeyUsage {
+				if !slices.Contains(TcgDiceExtendedKeyUsages[:], eku.String()) {
+					unknownKeyUsages = append(unknownKeyUsages, eku.String())
+				}
+			}
+
+			if len(unknownKeyUsagesMap) == 0 {
+				cert.UnknownExtKeyUsage = []asn1.ObjectIdentifier{}
+			} else {
+				unknownKeyUsagesMap[cert.Subject.String()] = unknownKeyUsages
+			}
+		}
+	}
+	// The error details in this map will be logged
+	if len(unknownKeyUsagesMap) > 0 {
+		for certSubject, ext := range unknownKeyUsagesMap {
+			t.Errorf("[ERROR]: Certificate \"%s\" has unknown extended key usages \"%s\"", certSubject, ext)
+		}
+		t.Errorf("[ERROR]: Certificate chain validation will fail with non-empty unknown extended key usages list")
+	}
+}
+
+// A tcg-dice-Ueid extension MUST be added
+// This SHALL be populated by the LABEL input parameter to CertifyKey
+// The extension SHOULD be marked as critical
+func checkCertifyKeyTcgUeidExtension(t *testing.T, c *x509.Certificate, label []byte) {
+	t.Helper()
+
+	isFound := false
+	// Check UEID extension
+	for _, ext := range c.Extensions {
+		if ext.Id.Equal(OidExtensionTcgDiceUeid) {
+			isFound = true
+			if !ext.Critical {
+				t.Errorf("[ERROR]: tcg-dice-Ueid extension is NOT marked as CRITICAL")
+			}
+			var ueid TcgUeidExtension = TcgUeidExtension{}
+			_, err := asn1.Unmarshal(ext.Value, &ueid)
+			if err != nil {
+				t.Errorf("[ERROR]: Error encountered while unmarshalling value of UEID extension, %s", err.Error())
+			}
+
+			t.Logf("[LOG]: Value of UEID extension is %v", ueid.Ueid)
+			t.Logf("[LOG]: Value of Label passed as input parameter is %v", label)
+			if !reflect.DeepEqual(ueid.Ueid, label) {
+				// Ueid extn value doen not match the label
+				t.Errorf("[ERROR]: tcg-dice-Ueid value does not match with the \"Label\" passed in CertifyKeyRequest")
+			} else {
+				t.Logf("[LOG]: tcg-dice-Ueid value matches with the \"Label\" passed in CertifyKeyRequest")
+			}
+			break
+		}
+	}
+	if !isFound {
+		t.Errorf("[ERROR]: tcg-dice-Ueid extension is missing")
+	}
+}
+
+// A tcg-dice-MultiTcbInfo extension.
+// This extension SHOULD be marked as critical.
+func checkCertifyKeyMultiTcbInfoExtension(t *testing.T, c *x509.Certificate) {
+	t.Helper()
+
+	// Check MultiTcbInfo Extension
+	//tcg-dice-MultiTcbInfo extension
+	for _, ext := range c.Extensions {
+		if ext.Id.Equal(OidExtensionTcgDiceMultiTcbInfo) { // OID for Tcg Dice MultiTcbInfo
+			if !ext.Critical {
+				t.Errorf("[ERROR]: TCG DICE MultiTcbInfo extension is not marked as CRITICAL")
+			}
+			multiTcbInfo, err := parseMultiTcbInfo(ext.Value)
+			if err != nil {
+				// multiTcb info is not provided in leaf
+				t.Errorf("[ERROR]: Failed to unmarshal MultiTcbInfo field: %v", err)
+			} else {
+				t.Logf("[LOG]: MultiTcb extension %v", multiTcbInfo)
+			}
+			break
+		}
+	}
+}
+
+func parseMultiTcbInfo(der []byte) (*TcgMultiTcbInfo, error) {
```

Comment:
We should avoid manually parsing ASN.1. The Go ASN.1 package provides type annotations, for example

```go
type diceTCBInfo struct {
	Vendor     string           `asn1:"optional,implicit,tag:0,utf8"`
	Model      string           `asn1:"optional,implicit,tag:1,utf8"`
	Version    string           `asn1:"optional,implicit,tag:2,utf8"`
	SVN        int              `asn1:"optional,implicit,tag:3"`
	Layer      int              `asn1:"optional,implicit,tag:4"`
	Index      int              `asn1:"optional,implicit,tag:5"`
	FWIDs      []fwid           `asn1:"optional,implicit,tag:6"`
	Flags      operationalFlags `asn1:"optional,implicit,tag:7"`
	VendorInfo []byte           `asn1:"optional,implicit,tag:8"`
	Type       []byte           `asn1:"optional,implicit,tag:9"`
}

type multiTCBInfo struct {
	DICETCBInfoSeq []diceTCBInfo
}
```

Then you can parse it with `asn1.Unmarshal()`

Reply:
I started out with this method. I works great if the data is not nested and has universal tags. 
Since we have some context specific tags, and it always resulted in tag mismatch errors. I even debugged into asn1 lib to understand and fix this. 
I came across some blogs suggesting that x509.parsecertificate that uses cryptobyte library is more efficient.
https://cs.opensource.google/go/go/+/refs/tags/go1.21.1:src/crypto/x509/parser.go;l=997 
Cryptobyte seemed more promising to deal with arrays and context-specific data types 
So switched to it. 

I see that using crypto byte provides more control to specify constructed, context specific tags.

Reply:
Do you still think this needs to be done with struct tags?

Reply:
The struct tagging mechanism supports implicit/explicit tags. I don't see what here can't be achieved with annotated structs.

The manual parsing code is a very large amount of code to maintain. I would greatly prefer the struct annotations.

Reply:
Hi Jordan, I have updated my code to rely only on struct tags.
It works fine now.
The issue came while defining the Multi Tcb Info improperly, it should have been an array not a struct it seems. 
Now it seems to be fixed. Thanks for insisting on this.